src/slurm_plugin/console_logger.py

```diff
@@ -0,0 +1,82 @@
+import logging
+import re
+from datetime import datetime, timezone
+
+import boto3
+from common.utils import sleep_remaining_loop_time
+
+logger = logging.getLogger(__name__)
+
+
+class ConsoleLogger:
+    """Class for retrieving and logging instance console output."""
+
+    def __init__(self, enabled, region, console_output_consumer, max_sample_size, console_output_wait_time):
+        self._region = region
+        self._console_logging_enabled = enabled
+        self._console_output_consumer = console_output_consumer
+        self._max_sample_size = max_sample_size
+        self._console_output_wait_time = console_output_wait_time
+        self._boto3_client_factory = lambda service_name: boto3.session.Session().client(
+            service_name, region_name=region
+        )
+
+    def report_console_output_from_nodes(self, compute_nodes, instance_supplier, task_executor):
+        """
+        Queue a task that will retrieve the console output for failed compute nodes.
+
+        The console output of up to self._max_sample_size nodes that have backing instances in `compute_nodes` will be
+        retrieved.
+        """
+        if not self._console_logging_enabled:
+            return None
+
+        start_time = datetime.now(tz=timezone.utc)
+
+        instances = tuple(instance for instance in instance_supplier(compute_nodes) if instance.get("InstanceId"))
+
+        if self._max_sample_size > 0:
+            instances = instances[: self._max_sample_size]
```

Comment:
nit: being very picky here but in case we want to spare the extra memory (and eventually some extra calls) we could just read max_sample_size instances from the supplier rather than reading all of them and then grabbing only the first max_sample_size.

Reply:
Refactored to do this.